File: backend/internal/model/item.go
// Package model contains request and response DTOs.
package model

import "time"

// CreateItemRequest defines the payload for creating a new item listing.
type CreateItemRequest struct {
	AddressID   string   `json:"address_id"    binding:"required,uuid"`
	Category    string   `json:"category"      binding:"required"`
	Title       string   `json:"title"         binding:"required,min=1,max=200"`
	Description string   `json:"description"   binding:"omitempty,max=2000"`
	UsageRules  string   `json:"usage_rules"   binding:"omitempty,max=2000"`
	Condition   string   `json:"condition"     binding:"required"`
	PricePerDay float64  `json:"price_per_day" binding:"required,gt=0"`
	Deposit     *float64 `json:"deposit"       binding:"omitempty,gte=0"`
	MinDays     int      `json:"min_days"      binding:"omitempty,gte=1"`
	MaxDays     *int     `json:"max_days"      binding:"omitempty,gte=1"`
}

// ItemResponse is the API representation of an item listing.
type ItemResponse struct {
	ItemID      string    `json:"item_id"`
	OwnerID     string    `json:"owner_id"`
	AddressID   string    `json:"address_id"`
	Category    string    `json:"category"`
	Title       string    `json:"title"`
	Description string    `json:"description,omitempty"`
	UsageRules  string    `json:"usage_rules,omitempty"`
	Condition   string    `json:"condition"`
	ItemStatus  string    `json:"item_status"`
	PricePerDay float64   `json:"price_per_day"`
	Deposit     *float64  `json:"deposit,omitempty"`
	MinDays     int32     `json:"min_days"`
	MaxDays     *int32    `json:"max_days,omitempty"`
	IsAvailable bool      `json:"is_available"`
	PublishedAt time.Time `json:"published_at"`
}

// SearchItemResponse is the compact item shape used by the search results UI.
type SearchItemResponse struct {
	ItemID          string    `json:"item_id"`
	OwnerID         string    `json:"owner_id"`
	AddressID       string    `json:"address_id"`
	Category        string    `json:"category"`
	Title           string    `json:"title"`
	ItemStatus      string    `json:"item_status"`
	PricePerDay     float64   `json:"price_per_day"`
	IsAvailable     bool      `json:"is_available"`
	PublishedAt     time.Time `json:"published_at"`
	City            string    `json:"city"`
	PrimaryImageURL string    `json:"primary_image_url,omitempty"`
}

// SearchItemsResponse wraps paginated search results.
type SearchItemsResponse struct {
	Items           []SearchItemResponse `json:"items"`
	Total           int64                `json:"total"`
	Page            int                  `json:"page"`
	Limit           int                  `json:"limit"`
	CategoryCounts  map[string]int64     `json:"category_counts"`
	CityCounts      map[string]int64     `json:"city_counts"`
	ConditionCounts map[string]int64     `json:"condition_counts"`
}

File: frontend/src/types/item.ts
/**
 * UI-level form state for the product creation form.
 * Not the same as the API payload — includes extra UI-only fields.
 */
interface ProductFormData {
  // Basic info
  title: string;
  category: string;
  subcategory: string;
  condition: string;
  description: string;
  // Photos
  photos: File[];
  // Price
  pricePerDay: string;
  pricePerWeek: string;
  deposit: string;
  minRentalPeriod: string;
  maxRentalPeriod: string;
  // Location
  address: string;
  city: string;
  deliveryRadius: string;
  availableNow: boolean;
  blockDates: boolean;
  // Conditions
  usageRules: string;
}

/**
 * Payload sent to POST /api/items.
 * Matches model.CreateItemRequest in the backend.
 */
interface CreateItemRequest {
  address_id: string;
  category: string;
  title: string;
  description?: string;
  usage_rules?: string;
  condition: string;
  price_per_day: number;
  deposit?: number;
  min_days: number;
  max_days?: number;
}

/**
 * API response for a single item listing.
 * Matches model.ItemResponse in the backend.
 */
interface ItemResponse {
  item_id: string;
  owner_id: string;
  address_id: string;
  category: string;
  title: string;
  description?: string;
  usage_rules?: string;
  condition: string;
  item_status: string;
  price_per_day: number;
  deposit?: number;
  min_days: number;
  max_days?: number | null;
  is_available: boolean;
  published_at: string;
}

/**
 * Compact item response used by the search results grid.
 */
interface SearchItemResponse {
  item_id: string;
  owner_id: string;
  address_id: string;
  category: string;
  title: string;
  item_status: string;
  price_per_day: number;
  is_available: boolean;
  published_at: string;
  city: string;
  primary_image_url?: string;
  owner_first_name: string;
  owner_last_name: string;
  owner_avatar_url?: string;
}

/**
 * Paginated search response returned by GET /api/items.
 */
interface SearchItemsResponse {
  items: SearchItemResponse[];
  total: number;
  page: number;
  limit: number;
  category_counts: Record<string, number>;
  city_counts: Record<string, number>;
  condition_counts: Record<string, number>;
}

/**
 * API response for a single item image stored in Supabase and item_image table.
 * Matches model.ItemImageResponse in the backend.
 */
interface ItemImageResponse {
  image_id: string;
  item_id: string;
  image_url: string;
  display_order: number;
}

/**
 * Category option used in the creation form dropdowns.
 */
interface CategoryOption {
  value: string;
  label: string;
  subcategories: string[];
}

/**
 * Compact item card returned in GET /api/favorites.
 * Matches model.FavoriteItemResponse in the backend.
 */
interface FavoriteItemResponse {
  item_id: string;
  category: string;
  title: string;
  item_status: string;
  price_per_day: number;
  is_available: boolean;
  city: string;
  primary_image_url?: string;
  saved_at: string;
}

/**
 * Response envelope for GET /api/favorites.
 */
interface FavoritesResponse {
  items: FavoriteItemResponse[];
  total: number;
}

/**
 * Response envelope for GET /api/favorites/:id/check.
 */
interface FavoriteCheckResponse {
  is_favorite: boolean;
}

export type {
  CategoryOption,
  CreateItemRequest,
  FavoriteCheckResponse,
  FavoriteItemResponse,
  FavoritesResponse,
  ItemImageResponse,
  ItemResponse,
  ProductFormData,
  SearchItemResponse,
  SearchItemsResponse,
};
